src/ORM/DataObjectSchema.php

```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
     }
 
     /**
-     * Cache all indexes for the given class.
-     * Will do nothing if already cached
+     * Cache all indexes for the given class. Will do nothing if already cached.
      *
      * @param $class
      */
     protected function cacheDatabaseIndexes($class)
     {
-        if (array_key_exists($class, $this->databaseIndexes)) {
-            return;
+        if (!array_key_exists($class, $this->databaseIndexes)) {
+            $this->databaseIndexes[$class] = array_merge(
+                $this->defaultDatabaseIndexes($class),
+                $this->customDatabaseIndexes($class)
+            );
         }
+    }
+
+    /**
+     * Get "default" database indexable field types
+     *
+     * @param  string $class
+     * @return array
+     */
+    protected function defaultDatabaseIndexes($class)
+    {
         $indexes = [];
+        if (empty($this->defaultDatabaseIndexes[$class])) {
+            $this->defaultDatabaseIndexes[$class] = [];
+        }
 
-        // look for indexable field types
-        foreach ($this->databaseFields($class, false) as $field => $type) {
-            if ($type === 'ForeignKey' || $type === 'DBClassName') {
-                $indexes[$field] = [
-                    'type' => 'index',
-                    'columns' => [$field],
-                ];
+        // Ensure that fixed field types, e.g. ClassName can be indexed
+        $databaseFields = $this->databaseFields($class, false);
```

Comment:
Actually, I'm not sure we can do this just with config - `DBPolymorphicForeignKey` is not a FQCN. We could define unqualified class name keys in a config file somewhere but then they wouldn't be bound to a concrete class. Dynamic options like `$className::class` don't work.